File: 1096-brace-expansion-ii/1096-brace-expansion-ii.py
class Solution:
    def braceExpansionII(self, expression: str) -> List[str]:
        def dfs(exp):
            j = exp.find('}')
            if j == -1:
                seen.add(exp)
                return
            i = exp.rfind('{', 0, j - 1)
            prefix, subfix = exp[:i], exp[j+1:]
            for ch in exp[i+1:j].split(','):
                dfs(prefix + ch + subfix)
                
                
        seen = set()
        dfs(expression)
        return sorted(seen)

File: 1096-brace-expansion-ii/1096-brace-expansion-ii.js
/**
 * @param {string} expression
 * @return {string[]}
 */
var braceExpansionII = function(expression) {
  const set = new Set();
  
  const dfs = (exp) => {
    const j = exp.indexOf('}');
    if (j === -1) return set.add(exp);
    
    let i = j;
    while (exp[i] !== '{') --i;
    
    const prefix = exp.substring(0, i);
    const subfix = exp.substring(j + 1);
    const curr = exp.substring(i + 1, j).split(',');
    
    curr.forEach((ch) => dfs(prefix + ch + subfix));
  }
  
  dfs(expression);
  
  return [...set].sort();
};